documentation/doxygen.py

```diff
@@ -1992,39 +1992,69 @@ def parse_func(state: State, element: ET.Element):
 
-    # Friend functions have friend as type. That's just awful. COME ON.
-    if func.type.startswith('friend '):
-        func.type = func.type[7:]
-
     def is_identifier(a): return a == '_' or a.isalnum()
 
     # Extract function signature to prefix, suffix and various flags. Important
     # things affecting caller such as static or const (and rvalue overloads)
     # are put into signature prefix/suffix, other things to various is_*
     # properties.
-    if func.type == 'constexpr': # Constructors
-        func.type = ''
-        func.is_constexpr = True
-    elif func.type.startswith('constexpr'):
-        func.type = func.type[10:]
-        func.is_constexpr = True
-    # For some effing reason, when a constexpr function has decltype(auto)
-    # return type, Doxygen swaps the order of those two, causing the constexpr
-    # to be last. See the cpp_function_attributes test for a verification.
-    elif func.type.endswith('constexpr'):
-        func.type = func.type[:-10]
-        func.is_constexpr = True
-    else:
-        func.is_constexpr = False
-    # When 1.8.18 encounters `constexpr static`, it keeps the static there. For
-    # `static constexpr` it doesn't. In both cases the static="yes" is put
-    # there correctly. WHY DOXYGEN, WHY?!
-    if func.type.startswith('static'):
-        func.type = func.type[7:]
+    #
+    # First the prefix keywords - Doxygen has a habit of leaking attributes and
+    # other specifiers into the function's return type, and not necessarily
+    # in any consistent order (including swapping it with the actual type!)
+    func.is_constexpr = False
+    func.is_consteval = False
+    func.is_explicit = False
+    func.is_virtual = False
+    func.is_static = False
+    keywords_behaving_badly = (
+        'constexpr',
+        'consteval',
+        'explicit',
+        'static',
+        'friend',
+        'inline',
+        'virtual'
+    )
+    matched_bad_keyword = True
+    while matched_bad_keyword:
+        matched_bad_keyword = False
+        for kw in keywords_behaving_badly:
+            if func.type == kw: # constructors
+                matched_bad_keyword = True
+                func.type = ''
+            elif func.type.startswith(kw+' '):
+                matched_bad_keyword = True
+                func.type = func.type[len(kw):].strip()
+            elif func.type.endswith(' '+kw):
+                matched_bad_keyword = True
+                func.type = func.type[:len(kw)].strip()
+            else:
+                continue
+            if kw == 'constexpr':
+                func.is_constexpr = True
+            elif kw == 'consteval':
+                func.is_consteval = True
+            elif kw == 'explicit':
+                func.is_explicit = True
+            elif kw == 'virtual':
+                func.is_virtual = True
+            elif kw == 'static':
+                func.is_static = True
```

Comment:
Hm, maybe some 

```py
setattr(func, 'is_' + kw, True)
``` 

instead of the `if` cascade? Would make the coverage happier also :)

Reply:
D: python's whole "everything is a dict lol" vibe continues to surprise me

Reply:
for me it's always "hmm let's see if python has something builtin for what i'd use a *nasty* preprocessor hack in C++", and it usually does :D

Reply:
I've applied this change, and also used it to replace the initial `blah = False` stuff before the loop.

Worth noting that a side-effect of this is that the `func` object now has the following new `is_` members:
- `is_consteval`
- `is_static`
- `is_friend`
- `is_inline`

I've updated the `.rst` file to make note of `is_consteval`, but do you want to do the same for the others? If they weren't important enough to be exposed as attributes before, it's not likely they are now, so seem like an implementation detail that can go undocumented.

Reply:
Ah, took me a while to understand what you mean (sleep is good, heh). For the sake of not having accidental undocumented features people might start relying on (because they always do), I'd partition the keyword list into exposed and ignored attributes:

```py
    exposed_attribute_keywords = [
        'constexpr',
        'consteval',
        'explicit',
        'virtual'
    ]
    ignored_attribute_keywords = [
        # Included in func.prefix already
        'static',
        # Ignored
        'friend',
        # Ignored
        'inline',
    ]

    for kw in exposed_attribute_keywords:
        setattr(func, 'is_' + kw, False)
        
    while matched_bad_keyword:
        ...
        for kw in exposed_attribute_keywords + ignored_attribute_keywords:
            ....

        if kw in exposed_attribute_keywords:
           setattr(func, 'is_' + kw, True)
```

Reply:
Ah, yes, the indomitable [Hyrum's Law](https://blog.axway.com/amplify-products/api-management/hyrums-law-and-api-design-and-management). Good point.

Somewhat complicates the 'merging' bit down below, but I'll figure something out ^_^

Reply:
Oh, never mind, it doesn't complicate it at all, lmao.

Reply:
Right, well, that's done. Can't help but feel the separation of the two sets of keywords does make it quite a bit messier, particularly factoring in the special handling of `static`. Would there be any great harm to just making them all publically-documented attributes, even if they're not used in the default doxygen theme? I can imagine scenarios where the additional context would be useful in custom themes.

Reply:
Ah, yes, the `is_static` might be considered a bit messy, but I think it's good enough now ... merging :)

I don't want to have one piece of information exposed multiple times, because if there's more than one way to express something, the two parts inevitably get out of sync. But of course if `is_static` would eventually make sense at some point in the future, sure (and then probably dropping the `prefix` in favor of it).